src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp

```diff
@@ -0,0 +1,241 @@
+  double travelCost = parameterAsDouble( parameters, QStringLiteral( "TRAVEL_COST" ), context );
+
+  bool includeBounds = true;  // default to true to maintain 3.0 API
+  if ( parameters.contains( QStringLiteral( "INCLUDE_BOUNDS" ) ) )
+  {
+    includeBounds = parameterAsBool( parameters, QStringLiteral( "INCLUDE_BOUNDS" ), context );
+  }
+
+  feedback->pushInfo( QObject::tr( "Building graph…" ) );
+  QVector< QgsPointXY > snappedPoints;
+  mDirector->makeGraph( mBuilder.get(), QVector< QgsPointXY >() << startPoint, snappedPoints, feedback );
+
+  feedback->pushInfo( QObject::tr( "Calculating service area…" ) );
+  QgsGraph *graph = mBuilder->graph();
+  int idxStart = graph->findVertex( snappedPoints[0] );
+
+  QVector< int > tree;
+  QVector< double > costs;
+  QgsGraphAnalyzer::dijkstra( graph, idxStart, 0, &tree, &costs );
+
+  QgsMultiPointXY points;
+  QgsMultiPolylineXY lines;
+  QSet< int > vertices;
+
+  int inboundEdgeIndex;
+  double startVertexCost, endVertexCost;
+  QgsPointXY edgeStart, edgeEnd;
+  QgsGraphEdge edge;
+
+  for ( int i = 0; i < costs.size(); i++ )
+  {
+    inboundEdgeIndex = tree.at( i );
+    if ( inboundEdgeIndex == -1 and i != idxStart )
+    {
+      // unreachable vertex
+      continue;
+    }
+
+    startVertexCost = costs.at( i );
+    if ( startVertexCost > travelCost )
+    {
+      // vertex is too expensive, discard
+      continue;
+    }
+
+    vertices.insert( i );
+    edgeStart = graph->vertex( i ).point();
+
+    // find all edges coming from this vertex
+    for ( int edgeId : graph->vertex( i ).outgoingEdges() )
+    {
+      edge = graph->edge( edgeId );
+      endVertexCost = startVertexCost + edge.cost( 0 ).toDouble();
+      edgeEnd = graph->vertex( edge.toVertex() ).point();
+      if ( endVertexCost <= travelCost )
+      {
+        // end vertex is cheap enough to include
+        vertices.insert( edge.toVertex() );
+        lines.push_back( QgsPolylineXY() << edgeStart << edgeEnd );
+      }
+      else
+      {
+        // travelCost sits somewhere on this edge, interpolate position
+        QgsPointXY interpolatedEndPoint = QgsGeometryUtils::interpolatePointOnLineByValue( edgeStart.x(), edgeStart.y(), startVertexCost,
+                                          edgeEnd.x(), edgeEnd.y(), endVertexCost, travelCost );
+
+        points.push_back( interpolatedEndPoint );
+        lines.push_back( QgsPolylineXY() << edgeStart << interpolatedEndPoint );
+      }
+    } // edges
+  } // costs
+
+  // convert to list so he have same order of points between algorithm runs
+  QList< int > verticesList = vertices.toList();
+  std::sort( verticesList.begin(), verticesList.end() );
+  for ( int v : verticesList )
+  {
+    points.push_back( graph->vertex( v ).point() );
+  }
+  // this produces correct results, but order of point is not the same, so tests failed
+  //QSet< int >::const_iterator it = vertices.constBegin();
```

Comment:
if you set the 'pk' compare option in the test yaml, does it help with this? That's designed to allow algorithms which export features in random orders.

Reply:
Ouch - that is tricky.